File: yaf-gadomain/src/main/java/ar/edu/ungs/yamiko/ga/domain/impl/BitSetGenome.java
package ar.edu.ungs.yamiko.ga.domain.impl;

import java.util.BitSet;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import ar.edu.ungs.yamiko.ga.domain.Gene;
import ar.edu.ungs.yamiko.ga.domain.Genome;
import ar.edu.ungs.yamiko.ga.domain.Ribosome;
import ar.edu.ungs.yamiko.ga.toolkit.StaticHelper;

/**
 * Implementación de Genoma orientado a tiras de bits (BitSet).
 * @author ricardo
 *
 */
public class BitSetGenome implements Genome<BitSet>{

	private Map<String,List<Gene>> structure;
	private int size;
	private Map<Gene,Ribosome<BitSet>> translators;

	public BitSetGenome(String uniqueCrhomosomeName,List<Gene> genes, Map<Gene,Ribosome<BitSet>> _translators) 
	{
		structure=new HashMap<String, List<Gene>>();
		structure.put(uniqueCrhomosomeName,genes);
		size=StaticHelper.calcGeneSize(structure.get(uniqueCrhomosomeName));
		translators=_translators;
	}
	
	public Map<String, List<Gene>> getStructure() {
		return structure;
		
		}
	
	public int size() {
		return size;
	}
		
	public Map<Gene,Ribosome<BitSet>> getTranslators() {
			return translators;
		}
}

File: yaf-gadomain-scala/src/main/scala/ar/edu/ungs/yamiko/ga/domain/impl/BitSetGenome.scala
package ar.edu.ungs.yamiko.ga.domain.impl

import ar.edu.ungs.yamiko.ga.domain.Gene
import ar.edu.ungs.yamiko.ga.domain.Genome
import ar.edu.ungs.yamiko.ga.domain.Ribosome
import scala.collection.mutable.BitSet


@SerialVersionUID(2159L)
class BitSetGenome(uniqueCrhomosomeName:String, genes:List[Gene], translators:Map[Gene,Ribosome[BitSet]]) extends Genome[BitSet]{
  
  //Constructor
	val structure:Map[String, List[Gene]]=Map(uniqueCrhomosomeName -> genes)
	val _size=genes.map(_.size()).sum
  
  override def size():Int=_size
  override def getStructure()=structure;
  override def getTranslators() =translators
    
	override def toString = "BitSetGenome [size=" + size + ", uniqueCrhomosomeName=" + uniqueCrhomosomeName + "]"
	
  def canEqual(a: Any) = a.isInstanceOf[BitSetGenome]

  override def equals(that: Any): Boolean =
    that match {
      case that: BitSetGenome => 
        that.canEqual(this) && this.hashCode == that.hashCode
      case _ => false
  }

  override def hashCode:Int = {
    val ourHash = if (uniqueCrhomosomeName == null) 0 else uniqueCrhomosomeName.hashCode+size.hashCode()    
    super.hashCode + ourHash
  }
  
}